### ISSUE TITLE
Cannot access member in descriptor for some type annotations

### ISSUE THREAD

**Issue by maurosilber** [microsoft/pyright#5535]
**Describe the bug**

Pylance reports that I cannot access a member in a descriptor.

**Code**

This works fine:

```python
from __future__ import annotations

class A:
    value: A | None = None

    def __get__(self, obj, cls):
        return self.value  # issue here when adding some type annotations to obj or cls

class B(A):
    pass
```

But adding some type annotation to `obj` or `cls`, such as

- `obj: B | None`
- `cls: type[B]`
- `obj: None`

produces `Cannot access member "value" for type "A"` in the line commented above.

I think it only works correctly if either there are no annotations in `__get__`, or `obj` or `cls` include `A` as the possible types (`obj: A`, `cls: type[A]`, or both, unioned with any other types).

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, or the command-line tool? Which version?

```
VS Code 1.80.1
pylance v2023.7.20
```

**Comment by erictraut** [comment]
Pyright is correct in generating an error if you use `B` in the annotation of `A.__get__`. 

The proper annotations are:
```python
    def __get__(self, obj: A | None, cls: type[A]):
        return self.value
```

or
```python
    def __get__(self, obj: Self | None, cls: type[Self]):
        return self.value
```

Not surprisingly, mypy generates the same error as pyright in this case.

**Comment by maurosilber** [comment]
Why? Running the following code I get:

```python
class A:
    def __get__(self, obj, cls):
        print(obj)

class B:
    x = A()

B.x  # prints `None`
B().x  # prints `<__main__.B object at 0x...>`
```

So, `obj` could be `B | None`. It depends on the classes where `A` is instanced, doesn't it?

**Comment by erictraut** [comment]
It _could_ be `B`, but it is not guaranteed to be. It could be an instance of `A` or some other subtype of `A` that is incompatible with `B`.

**Comment by maurosilber** [comment]
Sorry for the insistence, but the annotations for `property.__get__` in [typeshed](https://github.com/python/typeshed/blob/60a95da1592ca2950f6a87fc3901f25d737f9ee9/stdlib/builtins.pyi#L1227-L1244C1) use `Any`:

```python
class property:
    def __get__(self, __instance: Any, __owner: type | None = None) -> Any: ...
```

I don't see when `obj`, the instance of class on which the descriptor was instanced, would be `Self`. It would have to be something like:

```python
class Descriptor:
    def __get__(self, obj, cls): ...

    x = Descriptor()  # error here but...

Descriptor.x  # would call Descriptor.x.__get__ and cls would be Descriptor, if this were possible
```

So, I think `Any` is correct for `property`.

But I would expect typecheckers to understand that I intend to instance my descriptor on the annotated class for `obj`.

By the way, the issue seems to be only with class variables in a descriptor, as the instance variable in the following example does not show any error, in spite of the annotation on `obj` and `cls`:

```python
from __future__ import annotations

class A:
    class_var: A | None = None

    def __init__(self, value: A | None = None):
        self.inst_var = value

    def __get__(self, obj: B | None, cls: type[B]):
        self.inst_var  # ok, type is A | None
        self.class_var  # error, cannot access member "class var"

class B:
    x = A()
```

**Comment by erictraut** [comment]
Perhaps it will become clearer if I rewrite your original sample as follows:

```python
from __future__ import annotations

class A:
    value: A | None

    def __get__(self, obj: B | None, cls: type[B]):
        print("__get__ was called with", obj)
        assert obj.__class__ is A
        return 0

class B(A):
    x = A()

a = A()
A.value = A()
print(a.value)
```

When you run this code in the interpreter, the assert does not fire. That confirms that `obj` is an instance of `A`. Clearly, there was a type violation here because `obj` was annotated as `B | None`, but `A` is not a subtype of `B | None`. So it would be surprising if a type checker did not emit an error somewhere in this code. And indeed, pyright and mypy both catch this type violation. They report it when you attempt to access `a.value` because that's when the `__get__` method is implicitly invoked.

Looking back at your original code sample, you are invoking `self.value`, but this time it's from within the `__get__` method, and `self` is of type `Self@A` which is effectively `A` for the purposes of type checking.

It's very unusual for a descriptor class to have a class variable that has an attribute that refers to an instance of itself, as you're doing with `value` in the top sample. Is that really what you intend?

In the most recent sample above, you've annotated `class_var` to accept either `A` or `None`. `A` is a descriptor class. When you access a class variable that contains a descriptor, the descriptor's `__get__` method is invoked. When you access an instance variable that contains a descriptor, the descriptor's `__get__` method is not invoked. That's why pyright treats `self.inst_var` and `self.class_var` differently. It's simply modeling the runtime behavior.
